### ISSUE TITLE
The implenentation of cross-entropy loss of domain adaptive ensemble learning

### ISSUE THREAD

**Issue by KongMingxi** [KaiyangZhou/Dassl.pytorch#13]
Hi,

Thanks for your code and this is a great work.

I have read the paper and as described in the section 3, the loss function for domain-specific expert learning is a cross-entropy loss. 

I guess the implementation of the loss of domain adaptive ensemble learning is as following:
[https://github.com/KaiyangZhou/Dassl.pytorch/blob/master/dassl/engine/dg/daeldg.py#L109](url)

Is this a standard cross-entropy loss function? Why not use nn.CrossEntropyLoss()?

And why calculate the mean when calculating the cross entropy of each data sample?
(-label_i * torch.log(pred_i + 1e-5)).sum(1).mean()

**Comment by KaiyangZhou** [comment]
> Is this a standard cross-entropy loss function? Why not use nn.CrossEntropyLoss()?

Yes, it is a standard cross-entropy loss. `nn.CrossEntropyLoss` doesn't fit because the input label is one-hot vector

> why calculate the mean when calculating the cross entropy of each data sample

It's common to use mean reduction. You can use `sum` if you like, it's the same, but you might need to tune the learning rate or other coefficients (like adding a scaling value) to make sure the gradient is not too large

**Comment by KongMingxi** [comment]
> > Is this a standard cross-entropy loss function? Why not use nn.CrossEntropyLoss()?
> 
> Yes, it is a standard cross-entropy loss. `nn.CrossEntropyLoss` doesn't fit because the input label is one-hot vector
> 
> > why calculate the mean when calculating the cross entropy of each data sample
> 
> It's common to use mean reduction. You can use `sum` if you like, it's the same, but you might need to tune the learning rate or other coefficients (like adding a scaling value) to make sure the gradient is not too large

I see. Many thanks.
